### PR TITLE
Snowflake agent

### PR DESCRIPTION
**Pull request by hhcs9527**
# TL;DR
Support Snowflake with an agent, let the user is able to run it either locally or remotely.

## Type
 - [ ] Bug Fix
 - [X] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [ ] Smoke tested
 - [x] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
Support Snowflake with an agent, and let the user be able to run it either locally or remotely.

Here is an example.
```python
try:
    from typing import Annotated
except ImportError:
    from typing_extensions import Annotated

import pandas as pd
from flytekit import StructuredDataset, kwtypes, task, workflow
from flytekitplugins.snowflake import SnowflakeConfig, SnowflakeTask
from flytekit.types.schema import FlyteSchema

# we need to type our account, and user, and prepare the private key to build a connection.
# Note that, the table field is for building the correct uri for future reference.

snowflake_task_templatized_query2 = SnowflakeTask(
    name="sf",
    # Define inputs as well as their types that can be used to customize the query.1
    inputs=kwtypes(number=int),
    output_schema_type=StructuredDataset ,
    task_config=SnowflakeConfig( user='hhcs9527',
    account='ed61997.europe-west4.gcp',
    database='AGENT',
    schema='PEOPLE',
    table='INTRO',
    warehouse='COMPUTE_WH'),
    query_template="SELECT AGE, NAME from INTRO limit %(number)s",
)

@task
def convert_bq_table_to_pandas_dataframe(sd: DogeCoinDataset) -> pd.DataFrame:
    return sd.open(pd.DataFrame).all()

@workflow
def sf_query_wf(number: int) -> pd.DataFrame:
    sd = snowflake_task_templatized_query2(number=number)
    return convert_bq_table_to_pandas_dataframe(sd=sd)

if __name__ == "__main__":
    print(sf_query_wf(number=2)
```

## Tracking Issue
[Flyte Agent Ecosystem](https://github.com/flyteorg/flyte/issues/3936)

## Follow-up issue
[Add encoder/decoder in structureDataset for snowflake.](https://github.com/flyteorg/flytekit/pull/1811)

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -75,6 +75,7 @@
         "kubernetes>=12.0.1",
         "rich",
         "rich_click",
+        "snowflake-connector-python>=3.1.0",
```
Comment (pingsutw) [inline setup.py:78]: We should add it to the `flyteplugins-snowflake/setup.py `

File: plugins/flytekit-snowflake/flytekitplugins/snowflake/agent.py [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+        self.cs = self.conn.cursor()
+        self.cs.execute_async(task_template.sql.statement, params=params)
+        metadata = Metadata(query_id=self.cs.sfqid)
+
+        return CreateTaskResponse(resource_meta=json.dumps(asdict(metadata)).encode("utf-8"))
+
+    def get(self, context: grpc.ServicerContext, resource_meta: bytes) -> GetTaskResponse:
+        metadata = Metadata(**json.loads(resource_meta.decode("utf-8")))
+        try:
+            query_status = self.conn.get_query_status_throw_if_error(metadata.query_id)
```
Comment (pingsutw) [inline plugins/flytekit-snowflake/flytekitplugins/snowflake/agent.py:84]: Agent is designed to be stateless, so here we should get a new snowflake client.

File: plugins/flytekit-snowflake/flytekitplugins/snowflake/agent.py [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+        res = None
+
+        if cur_state == SUCCEEDED:
+            ctx = FlyteContextManager.current_context()
+            self.cs.get_results_from_sfqid(metadata.query_id)
+            res = literals.LiteralMap(
+                {
+                    "results": TypeEngine.to_literal(
+                        ctx,
+                        StructuredDataset(dataframe=self.cs.fetch_pandas_all()),
```
Comment (pingsutw) [inline plugins/flytekit-snowflake/flytekitplugins/snowflake/agent.py:100]: We shouldn't read all the data here. Use `snowflake://`URI instead. We need to add a snowflake encoder/decoder in the structured dataset first.

File: plugins/flytekit-snowflake/flytekitplugins/snowflake/agent.py [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+            user=config["user"],
+            password=config["password"],
+            account=config["account"],
+            database=config["database"],
+            schema=config["schema"],
+            warehouse=config["warehouse"]
+        )
+
+        self.cs = self.conn.cursor()
+        self.cs.execute_async(task_template.sql.statement, params=params)
```
Comment (pingsutw) [inline plugins/flytekit-snowflake/flytekitplugins/snowflake/agent.py:76]: We should render the query. https://github.com/flyteorg/flyteplugins/blob/03f5b5c709277f49b04b65b591c469695defd8d4/go/tasks/plugins/webapi/snowflake/plugin.go#L88-L95

**Comment by hhcs9527** [comment]
Hi @pingsutw , just finish this pr and related test, please take a look!

File: plugins/flytekit-snowflake/flytekitplugins/snowflake/task.py [hunk 5]
```diff
@@ -74,12 +81,14 @@ def __init__(
         )
         self._output_schema_type = output_schema_type
 
-    def get_config(self, settings: SerializationSettings) -> Dict[str, str]:
+    def get_custom(self, settings: SerializationSettings) -> Dict[str, str]:
```
Comment (pingsutw) [inline plugins/flytekit-snowflake/flytekitplugins/snowflake/task.py:84]: why do we change it to `get_custom`?

Comment (hhcs9527) [reply plugins/flytekit-snowflake/flytekitplugins/snowflake/task.py:84]: I will change it back to config

**Comment by hhcs9527** [comment]
Hi @pingsutw , just finished the change of the code by changin custom to config, please approve the testing and take a look.

**Comment by hhcs9527** [comment]
Hi @pingsutw, please approve the CI test.

**Review by pingsutw — COMMENTED** [review]
Overall LGTM

File: plugins/flytekit-snowflake/flytekitplugins/snowflake/agent.py [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+    def __init__(self):
+        super().__init__(task_type=TASK_TYPE)
+
+    def get_private_key(self):
+        from cryptography.hazmat.backends import default_backend
+        from cryptography.hazmat.primitives import serialization
+
+        import flytekit
+
+        pk_path = flytekit.current_context().secrets.get_secrets_file(TASK_TYPE, "rsa_key.p8")
```
[the remark this answers is not in the document]
Comment (hhcs9527) [reply plugins/flytekit-snowflake/flytekitplugins/snowflake/agent.py:49]: sure, we can replace private_key instead of rsa_key.p8 here.
However, since we need to read the path of the private_key and do the process, we need to use `get_secrets_file`

**Comment by hhcs9527** [comment]
Hi @pingsutw, please approve the CI test, thanks

**Comment by hhcs9527** [comment]
Hi @pingsutw, just finished the changes, Please take a look.

**Comment by hhcs9527** [comment]
Hi @pingsutw, this PR has been fully tested.
Please help me review it, thanks!

**Comment by hhcs9527** [comment]
Hi @pingsutw, just finished the changes to get, please review.
Thanks a lot!

**Review by pingsutw — APPROVED** [review]
